engine/consensus/approvals/assignment_collector.go

```diff
@@ -0,0 +1,386 @@
+// AssignmentCollector is responsible for validating approvals on result-level(checking signature, identity).
+type AssignmentCollector struct {
+	ResultID                             flow.Identifier                        // ID of execution result
+	result                               *flow.ExecutionResult                  // execution result that we are collecting approvals for
+	BlockHeight                          uint64                                 // height of block targeted by execution result
+	collectors                           map[flow.Identifier]*ApprovalCollector // collectors is a mapping IncorporatedBlockID -> ApprovalCollector
+	authorizedApprovers                  map[flow.Identifier]*flow.Identity     // map of approvers pre-selected at block that is being sealed
+	incorporatedAtHeight                 map[uint64][]flow.Identifier           // mapping blockHeight -> []IncorporatedBlockID
+	lock                                 sync.RWMutex                           // lock for protecting collectors map
+	incorporatedAtHeightLock             sync.Mutex
+	verifiedApprovalsCache               *ApprovalsCache                 // in-memory cache of approvals (already verified)
+	requiredApprovalsForSealConstruction uint                            // number of approvals that are required for each chunk to be sealed
+	assigner                             module.ChunkAssigner            // used to build assignment
+	state                                protocol.State                  // used to access the  protocol state
+	verifier                             module.Verifier                 // used to validate result approvals
+	seals                                mempool.IncorporatedResultSeals // holds candidate seals for incorporated results that have acquired sufficient approvals; candidate seals are constructed  without consideration of the sealability of parent results
+	approvalConduit                      network.Conduit                 // used to request missing approvals from verification nodes
+	requestTracker                       *sealing.RequestTracker         // used to keep track of number of approval requests, and blackout periods, by chunk
+	getCachedBlockHeight                 GetCachedBlockHeight            // functor to get cached block height
+}
+
+func NewAssignmentCollector(result *flow.ExecutionResult, state protocol.State, assigner module.ChunkAssigner, seals mempool.IncorporatedResultSeals,
+	sigVerifier module.Verifier, approvalConduit network.Conduit, requestTracker *sealing.RequestTracker, getCachedBlockHeight GetCachedBlockHeight, requiredApprovalsForSealConstruction uint) (*AssignmentCollector, error) {
+	blockHeight, err := getCachedBlockHeight(result.BlockID)
+	if err != nil {
+		return nil, err
+	}
+
+	collector := &AssignmentCollector{
+		verifiedApprovalsCache:               NewApprovalsCache(1000),
```

Comment:
I think we can fairly well calculate the limit of this cache. In case we have more verification nodes, a cache size of 1000 might quickly get to small.
* In the extreme and, every verifier could approve every chunk. Hence, we could set the limit to `result.Chunks.Len() * len(authorizedVerifiersAtBlock(result.BlockID))`
* If we have to eject elements, we know that we are under a memory overflow attack 😱